File: backend/app/models/expense.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

ExpenseCategory = Literal["court_fee", "shuttlecock", "jersey", "other"]


class ExpenseCreate(BaseModel):
    expense_date: date
    category: ExpenseCategory
    category_other: str | None = None
    amount: float
    paid_by: UUID
    note: str | None = None


class ExpenseUpdate(BaseModel):
    expense_date: date | None = None
    category: ExpenseCategory | None = None
    category_other: str | None = None
    amount: float | None = None
    paid_by: UUID | None = None
    note: str | None = None


class Expense(BaseModel):
    id: UUID
    expense_date: date
    category: ExpenseCategory
    category_other: str | None = None
    amount: float
    paid_by: UUID
    receipt_url: str | None = None
    note: str | None = None
    created_by: UUID
    created_at: datetime


class MonthlyExpenseSummary(BaseModel):
    """month is "YYYY-MM" — a plain string, not a date, since it has no day."""

    month: str
    total_amount: float
    by_category: dict[str, float]
    expense_count: int

File: frontend/src/types/expense.ts
export type ExpenseCategory = 'court_fee' | 'shuttlecock' | 'jersey' | 'other'

/** Mirrors backend app/models/expense.py Expense. */
export interface Expense {
  id: string
  expense_date: string
  category: ExpenseCategory
  category_other: string | null
  amount: number
  paid_by: string
  receipt_url: string | null
  note: string | null
  is_paid: boolean
  paid_at: string | null
  created_by: string
  created_at: string
}

/** Mirrors backend app/models/expense.py MonthlyExpenseSummary. */
export interface MonthlyExpenseSummary {
  month: string
  total_amount: number
  /** The part already reimbursed to the admin who fronted it — money that has
   * actually left the club's account. `total_amount` includes expenses still
   * owed back, which are a liability, not a withdrawal. */
  paid_amount: number
  by_category: Partial<Record<ExpenseCategory, number>>
  expense_count: number
}
